Fix the /dev/crypto engine

levitte opened the pull request:
It was lacking a digest copy function

levitte committed:
crypto/engine/eng_devcrypto.c: add digest copy

Copying an EVP_MD_CTX, including the implementation local bits, is a
necessary operation.  In this case, though, it's the same as
initializing the local bits to be "copied to".

Fixes #7495

levitte committed:
crypto/engine/eng_devcrypto.c: new compilers are strict on prototypes

FdaSilvaYY commented on crypto/engine/eng_devcrypto.c:
```diff
@@ -472,20 +472,26 @@ static int digest_final(EVP_MD_CTX *ctx, unsigned char *md)
 
+static int digest_copy(EVP_MD_CTX *to, const EVP_MD_CTX *from)
+{
+    /* There isn't much for us to copy, the |to| nees to be initialized */
```
s/nees/need

t-j-h reviewed (commented):
cfd is the handle for the crypto device - and its name entirely makes sense in that context.
renaming it seems illogical to me - it is named for what it is - not for which ioctls you plan to use on the device

levitte commented:
> cfd is the handle for the crypto device

Ah ok.  I'll undo that commit then.

cotequeiroz commented:
Something isn't quite right. It doesn't crash like it did before, but now it fails:
```
LD_LIBRARY_PATH=. ./apps/openssl s_client -connect localhost:443
CONNECTED(0000001F)
depth=2 O = Digital Signature Trust Co., CN = DST Root CA X3
verify return:1
depth=1 C = US, O = Let's Encrypt, CN = Let's Encrypt Authority X3
verify error:num=7:certificate signature failure
139866900886528:error:1416D07B:SSL routines:tls_process_key_exchange:bad signature:ssl/statem/statem_clnt.c:2414:
139866900886528:error:02014009:system library:close:Bad file descriptor:crypto/engine/eng_devcrypto.c:495:
---
```

cotequeiroz commented:
By looking at the old cryptodev engine, it seems that copy does more than initialize the new session.  It copies the state of the source digest operation, which includes the current hash output.  I assume we must do the same.

levitte commented:
The new devcrypto engine is designed to use `COP_FLAG_UPDATE` and `COP_FLAG_FINAL`, so it's not holding a cache of its own, it's leaving that to the kernel module.  But of course, with this copy function, we need to reimplement (1) that cache, unless someone knows a way to have the kernel module copy its cache, or somehow get a cache out of it and insert it into another session.

-----
(1) I mean "reimplement" seriously.  The old cryptodev engine was taken out for relicensing disagreement and the new devcrypto engine was reimplemented from scratch without looking at the old, to avoid "poisoning".  So if we're to insert a cache mechanism, it needs to be reimplemented in the same spirit...

levitte commented:
(I want to lament, yet again, what seems to be an utter lack of documentation of the `/dev/crypto` functionality.  If someone has good documentation, *please share*)

cotequeiroz commented:
I'm a noob with the details of digest algorithms, but given the digest block size, can't we just call the digest function with the full blocks, and save just the partial data for later.  That way we just need to copy the current digest and the partial data.  I'm not sure how /dev/crypto update function works.  If we can pass the current digest, we just need to call update with the partial data, and we should have a copy of the ctx.

cotequeiroz commented:
Can we have an option (either compile time or configuration) to turn it off.  It seems to ignore  `default_algorithms=CIPHERS`.  I'd hate to see people triyng to use that on a router with 32MB of RAM.

levitte commented:
`default_algorithms=CIPHERS`?  I don't think the engines ever listened to that

cotequeiroz commented:
https://www.openssl.org/docs/manmaster/man5/config.html#Engine-Configuration-Module seems to indicate that.  The config gets read, but apparently it does not do much, at least not with the `openssl` app.  My first idea when I encountered this bug was to turn off digests with that, but it didn't work.

levitte commented:
Ah, now I understand what you're talking about.  I had forgotten about the config part...  How exactly did you write your config?

cotequeiroz commented:
I used the default openssl.cnf, just added, before any section:
```
openssl_conf=openssl_section
```
then
```
[openssl_section]
engines=engine_section

[engine_section]
devcrypto=devcrypto_section

[devcrypto_section]
default_algorithms = CIPHERS
```
I've setup a breakpoint at `int_def_cb()`, and it did what it was supposed to do, but the `openssl s_client` would still fail.

cotequeiroz commented:
What if I propose a "copy" function to cryptodev-linux: a `CIOCCPSESSION` ioctl, that would use linux `crypto_ahash_export`/`import` to do the copying.  It seems that the only supported platform for digests is cryptodev-linux anyway.  OpenBSD has killed /dev/crypto, and FreeBSD does not do update/final, so  digests are disabled.  I'll ask here first, because if you nak, then it won't be worth my effort.

levitte commented:
Sure, that sounds like a good idea. I hadn't even noticed the ahash export/import functions, good catch.

levitte commented:
cryptodev-linux/cryptodev-linux#41 noted.  I've adapted the code in this PR accordingly, would you mind trying it out?

cotequeiroz commented:
I haven't actually tried this, but I was working at both changes at the same, so I noticed you can't copy the state between different cfd's.  So my suggestion is to make the cfd global, and opening it during load, and closing it in unload.  My code was pretty close to yours, and it would fail when a copy of a copy was done.  When it tries to copy a session that was a copy of another sesion, we get a NULL `digest_from`.  Since this is an openssl-controlled context, I can't see cryptodev causing it--it would either fail ioctl, or generate bad output, none of which would make the digest_from pointer NULL.

cotequeiroz commented on crypto/engine/eng_devcrypto.c:
```diff
@@ -480,6 +480,28 @@ static int digest_final(EVP_MD_CTX *ctx, unsigned char *md)
+{
+    struct digest_ctx *digest_from =
+        (struct digest_ctx *)EVP_MD_CTX_md_data(from);
+    struct cphash_op cphash;
```
```suggestion
   struct digest_ctx *digest_to =
       (struct digest_ctx *)EVP_MD_CTX_md_data(to);
    struct cphash_op cphash;
```

cotequeiroz commented on crypto/engine/eng_devcrypto.c:
```diff
@@ -480,6 +480,28 @@ static int digest_final(EVP_MD_CTX *ctx, unsigned char *md)
+        return 0;
+    }
+
+    cphash.src_ses = from->sess.ses;
```
```suggestion
    cphash.src_ses = digest_from->sess.ses;
```

cotequeiroz commented on crypto/engine/eng_devcrypto.c:
```diff
@@ -480,6 +480,28 @@ static int digest_final(EVP_MD_CTX *ctx, unsigned char *md)
+    }
+
+    cphash.src_ses = from->sess.ses;
+    cphash.dst_ses = to->sess.ses;
```
```suggestion
    cphash.dst_ses = digest_to->sess.ses;
```

levitte replied:
Thanks.  Fixed.

cotequeiroz reviewed (changes requested):
`sess` is a member of `digest_ctx`, not `EVP_MD_CTX`.

levitte commented:
> I haven't actually tried this, but I was working at both changes at the same, so I noticed you can't copy the state between different cfd's.  So my suggestion is to make the cfd global, and opening it during load, and closing it in unload.

Ah, I was wondering about that.  Thanks for checking in on it.  So globalize `cfd` it is...

levitte committed:
crypto/engine/eng_devcrypto.c: open /dev/crypto only once

We opened /dev/crypto once for each session, which is quite unnecessary.
With this change, we open /dev/crypto once at engine init, and close
it on unload.

cotequeiroz commented:
I've used your code, added a debug message at every `hash` function printing the `EVP_MD_CTX` address, cfd (should probably drop it now that it is global), /dev/crypto session id, and some other parameters passed.  Then I tried to establish a connection to localhost:443. It fails.  Here's the full output:
http://termbin.com/kf6f

The part that caught my attention was, after the server certificates were validated, there's a call to digest_copy, using as the source, an EVP_MD_CTX that had called digest_cleanup before.  So its md_data was probably freed and set to NULL. Here are the life-cycles of the CTXs involved:
```
verify return:1
digest_init:     CTX=55cef3abb170, md_data=55cef3abbb30, Session=5d10f30c, digest=106
digest_update:   CTX=55cef3abb170, md_data=55cef3abbb30, Session=5d10f30c, count=133
digest_init:     CTX=55cef3abbc20, md_data=55cef3abbbe0, Session=322bd295, digest=106
digest_copy: src CTX=55cef3abb170, md_data=55cef3abbb30, Session=5d10f30c
             to  CTX=55cef3abbc20, md_data=55cef3abbbe0, Session=322bd295
digest_final:    CTX=55cef3abbc20, md_data=55cef3abbbe0, Session=322bd295
digest_cleanup:  CTX=55cef3abbc20, md_data=55cef3abbbe0, Session=322bd295
digest_cleanup:  CTX=55cef3abb170, md_data=55cef3abbb30, Session=5d10f30c
```
So `abbc20` was created as copy of `abb170` (SHA512), was finalized and cleanup.
After that, there were 6 SHA256 operations that I'm omitting, then there's the weird copy call:
```
digest_copy: src CTX=55cef3abbc20, md_data=00000000 is NULL
             to  CTX=55cef3abbb30, md_data=00000000 is NULL
```
Besides using a source CTX that was cleared before, the destination CTX (`abbb30`) was actually the md_data of the original `abb170` CTX.  This might be coincidence, since that memory area was freed, but we have a call to digest_copy being made with a source CTX that was freed (and not initialized again), and a destination that was never initialized.
http://termbin.com/9x5x shows another run, but the results are compatible.

cotequeiroz commented on crypto/engine/eng_devcrypto.c:
```diff
@@ -472,24 +455,40 @@ static int digest_final(EVP_MD_CTX *ctx, unsigned char *md)
+    struct digest_ctx *digest_to =
+        (struct digest_ctx *)EVP_MD_CTX_md_data(to);
+    struct cphash_op cphash;
+
```
```suggestion
    if (!EVP_MD_test_flags(to, EVP_MD_CTX_FLAG_NO_INIT)
        return 1;
```

levitte replied:
Fixed

cotequeiroz reviewed (changes requested):
If the `EVP_MD_CTX_FLAG_NO_INIT` flag is set, we shouldn't touch the the implementation specific data.  That explains why we were getting those NULL md_data.  However, this is not the only case where NULL md_data is involved.  Apparently this is legal, so we should test and just return 1.  I was always returning 0, and that's why it was failing.

cotequeiroz commented:
I think I have it working.  At least `openssl s_client` finally is.  Since I had to study this to some depth, I feel confident enough to add some features to this engine, such as:

-  using `CIOCGSESSINFO` to query if an algorithm is accelerated or not, and enabling, by default, only those that are;

- turn it into a dynamic engine

- add some configuration options, such as cipher/digest list, allow use of unaccelerated drivers.

- using the `EVP_MD_CTX_FLAG_ONESHOT` flag to save an ioctl call.

Is it something in which you would be interested?

levitte commented:
All that is interesting, possibly except making it a dynamic engine.  I'm uncertain about that part...  on the other hand, that should be a configurable detail (hmmmm, that'll have to be another PR).

Please create a separate PR for your enhancements.

cotequeiroz commented:
Opening a new PR was always my intent.  Thanks.
PS: What's wrong with making it dynamic? I always wondered why some of the engines were made dynamic, while others weren't.

levitte commented:
> PS: What's wrong with making it dynamic? I always wondered why some of the engines were made dynamic, while others weren't.

I've never really understood why we should make *built in* engines (*that we load by default on init*) dynamic.  I actually don't understand why we have the padlock and capi as dynamic engines for the same reason, but I wasn't the author.  It's not like they take up much memory anyway.

But like I said, this could be made a configurable option, selectable for each of our built in engines.

cotequeiroz commented on crypto/engine/eng_devcrypto.c:
```diff
@@ -472,24 +455,42 @@ static int digest_final(EVP_MD_CTX *ctx, unsigned char *md)
+
+    cphash.src_ses = digest_from->sess.ses;
+    cphash.dst_ses = digest_to->sess.ses;
+    return ioctl(cfd, CIOCCPHASH, &cphash);
```
```suggestion
    if (ioctl(cfd, CIOCCPHASH, &cphash) < 0) {
        SYSerr(SYS_F_IOCTL, errno);
        return 0;
    }
    return 1;
```
`ioctl` returns 0 on sucess and -1 on error, while `digest_copy` should return 0 on error and 1 on success.

levitte replied:
Fixed

cotequeiroz reviewed (changes requested):
I missed this little change the last time, sorry.  It should work after this.

cotequeiroz commented on crypto/engine/eng_devcrypto.c:
```diff
@@ -472,24 +455,46 @@ static int digest_final(EVP_MD_CTX *ctx, unsigned char *md)
+        SYSerr(SYS_F_IOCTL, errno);
         return 0;
     }
+    return ;
```
```suggestion
    return 1;
```

levitte replied:
Argh...  it's obviously getting late :wink:

cotequeiroz reviewed (changes requested):
Just another fixup.

cotequeiroz reviewed (approved):
It works now!
Cheers

levitte commented:
Cool, now we only need some other @openssl/committers to approve as well.

levitte commented:
commit history cleaned up, and rebased on fresher master.

levitte commented:
[ahem] maybe we should make sure that cryptodev-linux/cryptodev-linux#41 gets accepted too...  :wink:

bernd-edlinger commented on crypto/engine/eng_devcrypto.c:
```diff
@@ -28,6 +28,13 @@
+ * such as digest session data copying (see digest_copy()), but is also
+ * saner...  why re-open /dev/crypto for every session?
+ */
+int cfd;
```
why is this not static?

levitte replied:
Good question. I will fix tomorrow

levitte replied:
Fixed

bernd-edlinger commented on crypto/engine/eng_devcrypto.c:
```diff
@@ -598,9 +598,12 @@ static int devcrypto_digests(ENGINE *e, const EVP_MD **digest,
     destroy_all_digest_methods();
 #endif
+
+    close(cfd);
```
how can you be sure that cfd is initialized,
given that engine_load_devcrypto_int() returns void?

levitte replied:
The `ENGINE_add` call at the end of `engine_load_devcrypto_int` makes the difference.

bernd-edlinger replied:
Okay, but then we have a resource leak if the execution does not reach that point.

levitte replied:
Huh? Could you elaborate?

levitte replied:
I think I saw what you see.  Commit coming up!

levitte committed:
fixup! crypto/engine/eng_devcrypto.c: open /dev/crypto only once

bernd-edlinger commented on crypto/engine/eng_devcrypto.c:
```diff
@@ -611,14 +614,13 @@ void engine_load_devcrypto_int()
+#ifdef IMPLEMENT_DIGEST
     prepare_digest_methods();
 #endif
 
```
in the line below, if ENGINE_new fails, cfd is leaked, right?

levitte committed:
crypto/engine/eng_devcrypto.c: ensure we don't leak resources

If engine building fails for some reason, we must make sure to close
the /dev/crypto handle.

bernd-edlinger commented on crypto/engine/eng_devcrypto.c:
```diff
@@ -664,7 +666,7 @@ void engine_load_devcrypto_int()
+#ifdef IMPLEMENT_DIGEST
         || !ENGINE_set_digests(e, devcrypto_digests)
 #endif
         ) {
```
the error handling here is tricky, because if and only if the ENGINE_set_destroy_function(e, devcrypto_unload)
succeeded, the cfd will not leak

bernd-edlinger replied:
maybe move that up to the first if, close cfd only there?

bernd-edlinger commented on crypto/engine/eng_devcrypto.c:
```diff
@@ -624,8 +625,10 @@ void engine_load_devcrypto_int()
 #endif
 
-    if ((e = ENGINE_new()) == NULL)
+    if ((e = ENGINE_new()) == NULL) {
```
Okay, as is, but I would have done this:

```
if ((e = ENGINE_new()) == NULL
          || !ENGINE_set_destroy_function(e, devcrypto_unload))
      {
           close(cfd);
           return;
      }
```

levitte replied:
Ah!  Sure, I can do that.  Nice :-)

levitte committed:
fixup! crypto/engine/eng_devcrypto.c: ensure we don't leak resources

levitte commented:
cryptodev-linux/cryptodev-linux#41 has now been accepted without change, so it's safe to merge this as well (given approval)
